src/utils/indexer.hpp

```diff
@@ -0,0 +1,102 @@
+//========================================================================================
+// (C) (or copyright) 2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef UTILS_INDEXER_HPP_
+#define UTILS_INDEXER_HPP_
+
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+namespace parthenon {
+
+template <class... Ts>
+struct Indexer {
+  Indexer() : N{}, start{}, _size{} {};
+
+  explicit Indexer(std::pair<Ts, Ts>... Ns)
+      : N{GetFactors(std::make_tuple((Ns.second - Ns.first + 1)...),
+                     std::make_index_sequence<sizeof...(Ts)>())},
+        start{Ns.first...}, end{Ns.second...}, _size(((Ns.second - Ns.first + 1) * ...)) {
```

Comment:
did not know you could variatically unpack with `*`. That's cool.

Reply:
Yeah, fold expressions are available in c++17. I think you could do everything they do with template recursion, but fold expressions give much cleaner code.